javatests/arcs/android/storage/service/StorageServiceNgImplTest.kt

```diff
@@ -0,0 +1,245 @@
+  private val onProxyMessageCallback: suspend (StorageKey, UntypedProxyMessage) -> Unit =
+    { _: StorageKey, _: UntypedProxyMessage -> }
+  private val storageKey = RamDiskStorageKey("myCount")
+  private val messageCallback = object : IMessageCallback {
+    override fun onMessage(encodedMessage: ByteArray?) {}
+
+    override fun asBinder(): IBinder? { return null }
+  }
+  private val channelCallback = object : IStorageChannelCallback {
+    override fun onCreate(channel: IStorageChannel) {}
+
+    override fun asBinder(): IBinder? { return null }
+  }
+
+  @Before
+  fun setUp() {
+    StorageKeyParser.addParser(RamDiskStorageKey)
+  }
+
+  @Test
+  fun opensStorageChannel() = runBlockingTest {
+    val job = Job()
+    val channelCallback = object : IStorageChannelCallback {
+      override fun onCreate(channel: IStorageChannel?) {
+        assertThat(channel).isInstanceOf(StorageChannelImpl::class.java)
+        job.complete()
+      }
+      override fun asBinder(): IBinder? { return null }
+    }
+    val messageCallback = object : IMessageCallback {
+      override fun onMessage(encodedMessage: ByteArray?) {}
+
+      override fun asBinder(): IBinder? { return null }
+    }
+
+    val storageServiceNg = createStorageService(this)
+
+    val storeOptions = StoreOptions(storageKey, CountType())
+    val encodedStoreOptions = with(Parcel.obtain()) {
+      writeStoreOptions(storeOptions, ParcelableCrdtType.Count, 0)
+      marshall()
+    }
+
+    storageServiceNg.openStorageChannel(encodedStoreOptions, channelCallback, messageCallback)
+    job.join()
+  }
+
+  @Test
+  fun openingTwoChannels_forSameStore_resultsInTwoDistinctChannels() = runBlockingTest {
+    val deferredChannel1 = CompletableDeferred<IStorageChannel>()
+    val deferredChannel2 = CompletableDeferred<IStorageChannel>()
+
+    val storageServiceNg = createStorageService(this)
+
+    val storeOptions = StoreOptions(storageKey, CountType())
+    val encodedStoreOptions = with(Parcel.obtain()) {
+      writeStoreOptions(storeOptions, ParcelableCrdtType.Count, 0)
+      marshall()
+    }
+
+    storageServiceNg.openStorageChannel(
+      encodedStoreOptions,
+      createChannelCallback(deferredChannel1),
+      messageCallback
+    )
+    storageServiceNg.openStorageChannel(
+      encodedStoreOptions,
+      createChannelCallback(deferredChannel2),
+      messageCallback
+    )
+
+    val channel1 = deferredChannel1.await()
+    val channel2 = deferredChannel2.await()
+    assertThat(channel2).isNotSameInstanceAs(channel1)
+  }
+
+  @Test
+  fun invalidStoreOptionsFail() = runBlockingTest {
+    val storageServiceNg = createStorageService(this)
+    val e = assertFailsWith<AssertionError> {
+      storageServiceNg.openStorageChannel(
+        "fake store options".toByteArray(),
+        channelCallback,
+        messageCallback
+      )
+    }
+    assertThat(e).hasMessageThat()
+      .isEqualTo("ShadowParcel unable to unmarshall its custom format")
```

Comment:
Oof, this is not a great error message. `ShadowParcel` is a Robolectric thing, for testing purposes only. Essentially you're testing the error message from the test framework, rather than the real error message you'd get in an app. Even the AssertionError is probably not the same exception that you'd actually see.

You can just delete this test, I think. Sorry for leading you astray with that suggestion.

Reply:
Ok I've removed it.